### PR TITLE
Added `colorscale` to `axes.plot()`

### PR DESCRIPTION
**Pull request by alembcke**
## Overview: What does this pull request change?

Added `colorscale` and `colorscale_axis` to `axes.plot()` to allow for colorscales by value for line plots.

## Motivation and Explanation: Why and how do your changes improve the library?

Adds functionality to the library consistent with `Surface` and `OpenGLSurface`.

## Links to added or changed documentation pages

https://manimce--3148.org.readthedocs.build/en/3148/reference/manim.mobject.graphing.coordinate_systems.CoordinateSystem.html#manim.mobject.graphing.coordinate_systems.CoordinateSystem.plot

## Further Information and Comments

Example code:

```py
from manim import *

class PlotExample(Scene):
    def construct(self):
        # construct the axes
        axes = Axes(
            x_range=[0, 6],
            y_range=[-2, 2],
            tips=False,
        )

        curve = axes.plot(
                lambda x: 2 * np.sin(x), x_range=(0.0, 6, 0.001), colorscale=[BLUE, GREEN, YELLOW, ORANGE, RED]
        )

        self.add(axes, curve)
```

PlotExample_ManimCE_v0 17 2

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

File: manim/mobject/graphing/coordinate_systems.py [hunk 1]
```diff
@@ -601,6 +605,12 @@ def plot(
         use_vectorized
             Whether to pass in the generated t value array to the function. Only use this if your function supports it.
             Output should be a numpy array of shape ``[y_0, y_1, ...]``
+        colorscale
+            Colors of the function. Optional parameter used when coloring a function by values. Passing a list of colors
+            and a colorscale_axes will color the function by y-value. Passing a list of tuples in the form ``(color, pivot)``
```
Comment (behackl) [inline manim/mobject/graphing/coordinate_systems.py:610]:
```suggestion
            and a colorscale_axis will color the function by y-value. Passing a list of tuples in the form ``(color, pivot)``
```

File: manim/mobject/graphing/coordinate_systems.py [hunk 2]
```diff
@@ -588,6 +590,8 @@ def plot(
         function: Callable[[float], float],
         x_range: Sequence[float] | None = None,
         use_vectorized: bool = False,
+        colorscale: Union[Iterable[Color], Color] | None = None,
```
Comment (behackl) [inline manim/mobject/graphing/coordinate_systems.py:593]: Passing a single color, which appears to be allowed based on this type hint, yields
```
ValueError: invalid literal for int() with base 16: ''
```

Also, the second option (passing an interable containing tuples of the form `(color, pivot)` is not captured by this type hint, this should be fixed.

**Review by behackl — COMMENTED** [review]
Thanks for your contribution! I have noticed that a visually similar behavior can already be achieved by setting the color of the plot to a gradient, and then specifying the sheen direction. Here is what I mean:

```py
from manim import *

class PlotExample(Scene):
    def construct(self):
        # construct the axes
        axes = Axes(
            x_range=[0, 6],
            y_range=[-2, 2],
            tips=False,
        )

        curve = axes.plot(
                lambda x: 2 * np.sin(2*x),
                x_range=(0.0, 6, 0.001),
        )
        curve.set_color([BLUE, GREEN, YELLOW, ORANGE, RED])
        curve.set_sheen_direction(UP)

        self.add(axes, curve)
```
for setting the gradient along the y-axis, and the same code as above, but with a different sheen direction
```py
        curve.set_sheen_direction(RIGHT)
```
for the gradient along the x-axis.

I still like the proposed interface, because it resembles the interface for 3D plots -- however, it might be worth it to implement it using these simpler, and already accessible methods; just from a point of making maintaining a little bit easier. What do you think?

**Comment by alembcke** [comment]
What about the pivots?  To me those are key as I set values below 0 to be one color and above 0 to be another color.  Is it possible to set pivots with the sheen direction?

File: manim/mobject/graphing/coordinate_systems.py (hunk 2 above)
Comment (alembcke) [reply manim/mobject/graphing/coordinate_systems.py]: Is this the right way to do it?

```py
colorscale: Union[Iterable[Color], Iterable[Color, float]] | None = None,
```

**Comment by behackl** [comment]
> What about the pivots? To me those are key as I set values below 0 to be one color and above 0 to be another color. Is it possible to set pivots with the sheen direction?

True! I don't think our current implementation of gradients allows to manually set pivots. Maybe this is something that we could consider after we have #3020 and implement our own `ManimColorGradient` class... Until then, your "manual" implementation might be a more suitable way to go.

File: manim/mobject/graphing/coordinate_systems.py (hunk 2 above)
Comment (MrDiver) [reply manim/mobject/graphing/coordinate_systems.py]: https://github.com/ManimCommunity/manim/discussions/2822

**Comment by MrDiver** [comment]
It doesn't work with opengl renderer

**Comment by MrDiver** [comment]
Can i just say this is hellishly slow ?

**Comment by MrDiver** [comment]
It takes 155 seconds in cairo to draw this. I don't think this is acceptable tbh.
grafik

https://github.com/ManimCommunity/manim/assets/17803932/28b29753-4dd8-4ae4-9c67-8ad10504e16e

**Comment by MrDiver** [comment]
I would suggest a solution which somehow exploits `curve.set_stroke([BLUE, RED, GREEN, YELLOW])` because that might be the faster way to change the colors without making thousands of little mobjects when it's not needed.

**Comment by alembcke** [comment]
@MrDiver the number of sub-mobjects created is up to the user.  It comes from `x_range=(0.0, 6, 0.001)` in `axes.plot()`.  I choose `0.001` to ensure a nice smooth transition of colors.  But changing it to `0.1` still results in a decent transition of colors and takes a fraction of the time:

CurveExample_ManimCE_v0 17 3

Is that a reasonable solution?  It still doesn't work in OpenGL, but that as you point out is because `CurvesAsSubmobjects` doesn't work in OpenGL and so the appropriate solution would be to get `CurvesAsSubmobjects` working in OpenGL in a separate PR.

As for using `curve.set_stroke([BLUE, RED, GREEN, YELLOW])`, I don't see how that can be done while allowing to set at least a mid-point for the colors.  I say that as I often set certain colors for less than `0` and others for greater than `0`.  To me this is necessary and is already how `Surface` and `OpenGLSurface` work with colorscales (as well as other charting libraries like Plotly).  I think it is worth it to discuss how colorscales should be implemented across all three, as there is some duplication of code.  That is why I asked @behackl above about #3020 and implementing `ManimColorGradient`, as I would prefer that colorscales be implemented there and then each of these just use that implementation.

**Comment by MrDiver** [comment]
I already got the OpenGL thing to work, that's why i have the flamegraph for the 95 seconds. That's what it takes to render this I still think it is a little bit much to create so many mobjects for that. As far as i understand it colors may be applied independently of the number of mobjects.

**Comment by MrDiver** [comment]
@alembcke 
```py
class PlotExample(Scene):
    def construct(self):
        # construct the axes
        axes = Axes(
            x_range=[0, 6],
            y_range=[-2, 2],
            tips=False,
        )

        curve = axes.plot(
            lambda x: 2 * np.sin(x), x_range=(0.0, 6), use_vectorized=True
        )
        colors = [
            "#ff0000",
            "#ffa500",
            "#ffff00",
            "#008000",
            "#0000ff",
            "#4b0082",
            "#ee82ee",
        ]
        # curve.set_stroke(colors)
        curve.set_color(colors)
        # curve.set_fill(colors, opacity=1)
        self.play(Write(axes), Write(curve))
        self.play(curve.animate.set_sheen_direction(UP))
        self.play(curve.animate.set_sheen_direction(RIGHT))
        self.play(curve.animate.set_sheen_direction(LEFT))
        self.wait()
```

You might wanna take a look at this 👀 i just realized you can do that so it might make realizing the color stuff a lot easier because it is not dependend on the number of curves

https://github.com/ManimCommunity/manim/assets/17803932/fd0eb1b1-30f1-4cc6-ad4a-aaf2d8f96163

**Comment by MrDiver** [comment]
https://github.com/ManimCommunity/manim/assets/17803932/8cdf0038-ab4b-4f78-a19b-90e941f88f20

And you can scale the amount of colors without increasing the rendering time at all.

**Comment by MrDiver** [comment]
I added some changes to the PR, you can check if they have the same behavior that you intended. I just reused all of your code. I also added a test to check if the color axis parameter actually works so that it doesn't get missed in the graphical tests.

**Comment by alembcke** [comment]
Just to reiterate here what was discussed on Discord, I rendered a bunch of videos I have that use this functionality and they all behave the same way as before but render much faster.

**Review by MrDiver — APPROVED** [review]

**Review by MrDiver — CHANGES_REQUESTED** [review]
We need to review the Test data again because main changed something i suppose

**Comment by alembcke** [comment]
This is weird because the tests pass on my machine.  I even deleted the folder with Manim in it and did a `git clone` and reran the tests - they still pass.

I double and triple checked that I am on the `line_gradient` branch and I ran `poetry install` and `poetry run pytest`.  Am I doing anything wrong here?

**Review by MrDiver — APPROVED** [review]

**Pull request merged**